### PR TITLE
Prototype modifications made for RANZCR kaggle challenge  

### PR DESCRIPTION
**Pull request by aribornstein**
Here is the notebook for the competition https://www.kaggle.com/aaronbornstein/iterating-ranzcr-baselines-in-a-flash 

- Multi label support and one hot embedding
- Enable BCE loss with logits (regular BCE breaks with no warning)
- Support for binding and overridng functions
- Loading custom backbones for image classifcaiton  (should be applied to all models)
-Added support not to attach head to model backbone (should be applied to all models)
- prevent  redownloading image classification backbones when loading from checkpoint
- hardcoded support ablumenation transforms for image classification (needs to be cleaned)

Features would make things cleaner
-  LR schedulers - flash doesn’t support this
- Train with folds - flash doesn’t support this and should in it’s datamodule
-  Split data leakage check hook
- native support for timm backbones
- loading large models for inference is slow
- other open issues

### CODE REVIEW

**Comment by aribornstein** [comment]
> Hello @aribornstein! Thanks for opening this PR.
> 
> * In the file [`flash/core/classification.py`](https://github.com/PyTorchLightning/lightning-flash/blob/9df00f175b5cee8f47cf08f03c2fa58decd892b1/flash/core/classification.py):
> 
> > [Line 93:14](https://github.com/PyTorchLightning/lightning-flash/blob/9df00f175b5cee8f47cf08f03c2fa58decd892b1/flash/core/classification.py#L93): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> > [Line 95:44](https://github.com/PyTorchLightning/lightning-flash/blob/9df00f175b5cee8f47cf08f03c2fa58decd892b1/flash/core/classification.py#L95): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> > [Line 98:13](https://github.com/PyTorchLightning/lightning-flash/blob/9df00f175b5cee8f47cf08f03c2fa58decd892b1/flash/core/classification.py#L98): [E124](https://duckduckgo.com/?q=pep8%20E124) closing bracket does not match visual indentation
> > [Line 100:23](https://github.com/PyTorchLightning/lightning-flash/blob/9df00f175b5cee8f47cf08f03c2fa58decd892b1/flash/core/classification.py#L100): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> > [Line 101:23](https://github.com/PyTorchLightning/lightning-flash/blob/9df00f175b5cee8f47cf08f03c2fa58decd892b1/flash/core/classification.py#L101): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> > [Line 102:23](https://github.com/PyTorchLightning/lightning-flash/blob/9df00f175b5cee8f47cf08f03c2fa58decd892b1/flash/core/classification.py#L102): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> > [Line 103:23](https://github.com/PyTorchLightning/lightning-flash/blob/9df00f175b5cee8f47cf08f03c2fa58decd892b1/flash/core/classification.py#L103): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> 
> * In the file [`flash/core/utils.py`](https://github.com/PyTorchLightning/lightning-flash/blob/9df00f175b5cee8f47cf08f03c2fa58decd892b1/flash/core/utils.py):
> 
> > [Line 16:1](https://github.com/PyTorchLightning/lightning-flash/blob/9df00f175b5cee8f47cf08f03c2fa58decd892b1/flash/core/utils.py#L16): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> 
> * In the file [`flash/data/data_utils.py`](https://github.com/PyTorchLightning/lightning-flash/blob/9df00f175b5cee8f47cf08f03c2fa58decd892b1/flash/data/data_utils.py):
> 
> > [Line 35:1](https://github.com/PyTorchLightning/lightning-flash/blob/9df00f175b5cee8f47cf08f03c2fa58decd892b1/flash/data/data_utils.py#L35): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> > [Line 43:17](https://github.com/PyTorchLightning/lightning-flash/blob/9df00f175b5cee8f47cf08f03c2fa58decd892b1/flash/data/data_utils.py#L43): [E126](https://duckduckgo.com/?q=pep8%20E126) continuation line over-indented for hanging indent
> > [Line 43:71](https://github.com/PyTorchLightning/lightning-flash/blob/9df00f175b5cee8f47cf08f03c2fa58decd892b1/flash/data/data_utils.py#L43): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> 
> * In the file [`flash/vision/classification/model.py`](https://github.com/PyTorchLightning/lightning-flash/blob/9df00f175b5cee8f47cf08f03c2fa58decd892b1/flash/vision/classification/model.py):
> 
> > [Line 65:29](https://github.com/PyTorchLightning/lightning-flash/blob/9df00f175b5cee8f47cf08f03c2fa58decd892b1/flash/vision/classification/model.py#L65): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> > [Line 68:31](https://github.com/PyTorchLightning/lightning-flash/blob/9df00f175b5cee8f47cf08f03c2fa58decd892b1/flash/vision/classification/model.py#L68): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> > [Line 69:1](https://github.com/PyTorchLightning/lightning-flash/blob/9df00f175b5cee8f47cf08f03c2fa58decd892b1/flash/vision/classification/model.py#L69): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> 
> Do see the [Hitchhiker's guide to code style](https://goo.gl/hqbW4r)

@Borda Didn't we add an automatic linter for this?

File: flash/core/classification.py [hunk 1]
```diff
@@ -32,6 +35,79 @@ def after_uncollate(self, samples: Any) -> Any:
...
+    def predict(
+        self,
+        x: Any,
+        batch_idx: Optional[int] = None,
+        skip_collate_fn: bool = False,
+        dataloader_idx: Optional[int] = None,
+        data_pipeline: Optional[ClassificationDataPipeline] = None,
+    ) -> Any:    
+        if self.multilabel:
+            bind_method(self.data_pipeline,
```
Comment (tchaton) [inline flash/core/classification.py:95]: We shouldn't hard-code this kind of hack.

File: flash/vision/classification/data.py [hunk 2]
```diff
@@ -73,7 +75,13 @@ def __getitem__(self, index: int) -> Tuple[Any, Optional[int]]:
         filename = self.fnames[index]
         img = self.loader(filename)
         if self.transform is not None:
-            img = self.transform(img)
+            if str(type(self.transform)) == "<class 'albumentations.core.composition.Compose'>":
```
Comment (tchaton) [inline flash/vision/classification/data.py:78]: Not a huge fan of this hard-coded check. Any better solution ?

Comment (aribornstein) [reply flash/vision/classification/data.py:78]: Agreed this was developed just so that I could get unblocked as soon as possible. Remember this pull request is not meant to be merged it's rather meant to serve as a reference as potential solutions for the open issues that need to be resolved so that Flash can be used practically for Kaggle.

File: flash/vision/classification/model.py [hunk 3]
```diff
@@ -64,35 +43,50 @@ def __init__(
...
             loss_fn=loss_fn,
             optimizer=optimizer,
             metrics=metrics,
             learning_rate=learning_rate,
+            multilabel=multilabel,
         )
 
         self.save_hyperparameters()
 
+        # This is very hacky
```
Comment (aribornstein) [inline flash/vision/classification/model.py:65]: This prevents model from being downloaded when offline

File: flash/core/classification.py (hunk 1 above)
Comment (aribornstein) [reply flash/core/classification.py:95]: @tchaton @kaushikb11 I agree do we have anyone working on adding clean multilabel support 90% of the current Kaggle image classification challenges are multilabel so this is a pretty significant bottleneck towards flash adoption for Kaggle tasks.
